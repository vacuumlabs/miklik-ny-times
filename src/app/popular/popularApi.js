/**
 * Returns promise resolved to array of popular articles
 * 
 * @see ItemPropType
 */
export function getPopularArticles() {
  const url = `https://api.nytimes.com/svc/mostpopular/v2/viewed/7.json?api-key=${
    window.__NYMV_API_KEY
  }`
  return fetch(url)
    .then(response => response.json())
    .then(parseBody)
}

function parseBody(body) {
  if (body.status !== 'OK') {
    throw new Error(makeError(body))
  }
  return body.results.map(r => ({
    id: r.id,
    title: r.title,
    abstract: r.abstract,
    authors: r.byline,
    date: r.published_date,
    thumbnail: extractThumb(r.media),
  }))
}

function extractThumb(media = []) {
  const imgs = media.filter(m => m.type === 'image')
  return !!imgs.length ? imgs[0]['media-metadata'][0].url : undefined
}

function makeError(body) {
  let str = undefined
  if (body.fault) {
    str = body.fault.faultstring
  }
  return str || 'Request failed'
}
