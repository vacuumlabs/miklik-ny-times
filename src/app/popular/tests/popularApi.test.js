import { commonSetup } from '../../../tests/testUtils'
import { fullResponse, faultResponse } from '../../../tests/apiFixtures'
import { getPopularArticles } from '../popularApi'

describe('getPopularArticles', () => {
  commonSetup()

  it('properly extract response from API', async () => {
    fetch.mockResponseOnce(JSON.stringify(fullResponse))

    const response = await getPopularArticles()

    expect(response[0]).toMatchSnapshot()
  })

  it('extracts error from fault error', async () => {
    fetch.mockResponseOnce(JSON.stringify(faultResponse))

    let error = false
    try {
      await getPopularArticles()
    } catch (e) {
      error = true
      expect(e.message).toBe('API failed')
    }

    expect(error).toBe(true)
  })
})
