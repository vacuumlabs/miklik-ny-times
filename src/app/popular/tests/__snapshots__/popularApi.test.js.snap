// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPopularArticles properly extract response from API 1`] = `
Object {
  "abstract": "A high school valedictorian quoted Trump. The crowd roared. Then: “Just kidding,” he said. “That was Barack Obama.”",
  "authors": "By LAURA M. HOLSON",
  "date": "2018-06-04",
  "id": 100000005935794,
  "thumbnail": "https://static01.nyt.com/images/2018/06/04/us/05xp-graduation2/05xp-graduation2-thumbStandard.jpg",
  "title": "Kentucky Crowd Cheers Valedictorian’s Trump Quote, Then Learns Obama Said It",
}
`;
